sw/device/lib/dif/dif_pinmux.h

```diff
@@ -0,0 +1,656 @@
+ *
+ * A pad can be a Multiplexed (MIO) or Dedicated (DIO).
+ */
+typedef enum dif_pinmux_pad_kind {
+  /**
+   * Multiplexed Input/Output pad. Connected to a peripheral input or output
+   * via a multiplexer matrix inside the a Pin Multiplexer.
+   */
+  kDifPinmuxPadKindMio = 0,
+  /**
+   * Dedicated Input/Output pad. Connected directly to a peripheral input
+   * or output, bypassing the multiplexer matrix.
+   */
+  kDifPinmuxPadKindDio,
+} dif_pinmux_pad_kind_t;
+
+/**
+ * Pin Multiplexer functionality locking.
+ *
+ * Represents a target functionality to be locked for a given
+ * pad/peripheral input.
+ */
+typedef enum dif_pinmux_lock_target {
+  /**
+   * Lock MIO pad input select for a given peripheral input. This means that
+   * the peripheral input connection can no longer be changed.
+   */
+  kDifPinmuxLockTargetInsel = 0,
+  /**
+   * Lock peripheral output select for a given MIO pad output. This means
+   * that the MIO pad output connection can no longer be changed.
+   */
+  kDifPinmuxLockTargetOutsel,
+  /**
+   * Lock the sleep mode configuration for a given MIO pad.
+   */
+  kDifPinmuxLockTargetMioSleep,
+  /**
+   * Lock the sleep mode configuration for a given DIO pad.
+   */
+  kDifPinmuxLockTargetDioSleep,
+  /**
+   * Lock physical characteristics configuration for a given MIO pad.
+   */
+  kDifPinmuxLockTargetMioPadAttr,
+  /**
+   * Lock physical characteristics configuration for a given DIO pad.
+   */
+  kDifPinmuxLockTargetDioPadAttr,
+  /**
+   * Lock wake-up detector configuration for a given detector.
+   */
+  kDifPinmuxLockTargetWakeupDetector,
+
+} dif_pinmux_lock_target_t;
+
+/**
+ * Pin Multiplexer generic index.
+ *
+ * The meaning of this index changes between APIs, and must be documented in
+ * the corresponding function description.
+ */
+typedef uint32_t dif_pinmux_index_t;
+
+/**
+ * Pin multiplexer Padring pad attributes.
+ *
+ * This `enum` is a bitfield, meaning that it is allowed to pass combined value
+ * of multiple individual attributes.
+ *
+ * Some attributes have defined meanings, others are implementation defined. Not
+ * all pads will support all attributes, or attribute combinations. The Pin
+ * Multiplexer DIF does not check for conflicting or unsupported attributes,
+ * however the underlying hardware prevents these from being enabled. If the
+ * combination of attributes is not supported - some/all of these attributes
+ * will still be disabled (relevant bit unset in the attribute field).
+ *
+ * NOTE:
+ * The functions are used to toggle these attributes, must signal an error if at
+ * least one of the attributes in the bitfield could not be toggled.
+ */
+typedef enum dif_pinmux_pad_attr {
+  kDifPinmuxPadAttrNone = 0,
+  kDifPinmuxPadAttrIOInvert = 1 << 0,
+  kDifPinmuxPadAttrVirtOpenDrain = 1 << 1,
+  kDifPinmuxPadAttrPullDown = 1 << 2,
+  kDifPinmuxPadAttrPullUp = 1 << 3,
+  kDifPinmuxPadAttrKeeper = 1 << 4,
+  kDifPinmuxPadAttrWeakestDrive = 1 << 5,
```

Comment:
What do you do if no drive strength is specified? It has to be one of four options.

Reply:
Well, wouldn't it be caught by WARL?

Reply:
Nope, you're writing a bunch of bits to the attribute register, two of them control drive strength, all 4 possible options are valid. The API allows the user to not specify any option. When producing the bit field for the attribute register write the DIF has to choose one of them even if none was provided. If you didn't really consider this in the DIF this choice will likey be 0, which is a valid drive strength (weakest) so no WARL error would be seen.

Reply:
I see, well yeah, well I suppose it would then default to weakest. I could reflect that in the description.

Reply:
Do you think that documenting default choices would suffice?

Reply:
I think this comes back to my suggestion that having this enum more closely match the register values recognised by the hardware might simplify things overall.

Reply:
It doesn't solve the above, as you could still happily pass a bitfield that doesn't have any of the strength bits set, and the hardware will gladly accept it.

Reply:
Yes, sorry - my comment was incorrect.

I think in addition to the concern about someone not specifying the drive strength at all, Greg is making the point that drive strength is actually 00, 01, 10, or 11.

You could pull drive strength out to a separate parameter when setting the pad attributes (moving further away from my previous suggestion of more closely matching the register encodings!).

Reply:
Updated the description, it should now return an error if none of these compulsory attribute invariants are specified.

So really it's a combination of 1. and 2.

Splitting out some attributes into their own enums has proven to be cumbersome, especially when then checking `attrs_in` and `attrs_out`. I am keen on going with the 1. and 2., and come back to it if required during the implementation.

I am eager to merge this header by the end of this week to keep-up the progress. I feel like this header is in a good state.

Reply:
Yep, I'm eager to get this merged too. As you say, some minor details can be revisited once there's an implementation.

@msfschaffner - what do you think about the weakest/weak/strong/strongest naming for driving strength

Reply:
Our current pad library only has 2 driving strength bits, so four named enums are fine for now, and I think we can move on to make progress.

However, please keep in mind that other pad libraries may provide 3 or 4 bits for this attribute, see [this Raspberry Pi doc page](https://www.raspberrypi.org/documentation/hardware/raspberrypi/gpio/gpio_pads_control.md) for example.

We may thus want to consider making this either just a number, or provide more, numbered enum values (like `kDifPinmuxPadAttrDrive0`, `kDifPinmuxPadAttrDrive1` etc).
In order to make this future proof we can also widen the driving strength attribute width from 2 to 4 bits.

This would also simplify the implementation of the WARL behavior in HW.
I.e., if a pad only supports 2 strength levels, you would only be allowed to write `kDifPinmuxPadAttrDrive0`, `kDifPinmuxPadAttrDrive1`, encoded as 0 and 1.
This is easy to check since you would basically just disallow 1'es in the upper bit positions.

With the named enums this gets a bit more complicated since in the 2 strength level case you would likely want to use weakest/strongest, which would be encoded as 11 and 00 - so a per bit WARL check does not work and you have to decode the combinations.